### PR TITLE
sql: support copy from stdin format csv

### PR DESCRIPTION
**Pull request by sploiselle**
### Motivation

This PR adds a known-desirable feature, `COPY FROM ... (FORMAT CSV)` as part of Fivetran support https://github.com/MaterializeInc/materialize/issues/5188#issuecomment-759262331

### Checklist

- [x] This PR has adequate test coverage / QA involvement has been duly considered.
- [x] This PR adds a release note for any
  [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/user/content/release-notes.md#what-changes-require-a-release-note).

### CODE REVIEW

File: test/pgtest/copy-from-2.pt [hunk 1]
```diff
@@ -59,6 +61,7 @@ CopyIn {"format":"text","column_formats":["text","text"]}
 ErrorResponse {"fields":[{"typ":"C","value":"22P04"}]}
 ReadyForQuery {"status":"I"}
 
+
```
Comment (sploiselle) [inline test/pgtest/copy-from-2.pt:64]: Will fix this on next push.

**Comment by sploiselle** [comment]
@mjibson This now actually does something meaningful; ready for a re-review.

File: src/pgwire/src/message.rs [hunk 2]
```diff
@@ -796,6 +885,142 @@ pub fn decode_copy_text_format(
...
+        .quote(quote)
+        .has_headers(header)
+        .double_quote(double_quote)
+        .escape(escape)
+        // Must be flexible to accept end of copy marker, which will always be 1
+        // field.
+        .flexible(true)
+        .from_reader(data);
+
+    let records = rdr
```
Comment (mjibson) [inline src/pgwire/src/message.rs:977]: Is there a reason to pre-parse the whole thing instead of row-by-row?

File: src/pgwire/src/protocol.rs [hunk 3]
```diff
@@ -1534,13 +1539,8 @@ where
...
             .collect::<Vec<pgrepr::Type>>();
 
-        if let State::Ready = next_state {
-            let rows = match decode_copy_text_format(
-                &data,
-                &column_types,
-                &params.delimiter,
-                &params.null,
-            ) {
+        if next_state == State::Ready {
```
Comment (mjibson) [inline src/pgwire/src/protocol.rs:1542]: what's wrong with the Let? that'd allow you to not derive all the stuff on state above.

File: test/pgtest/copy-from-csv.pt [hunk 4]
```diff
@@ -0,0 +1,139 @@
+# Note that this file will not work on PG due to the lack of "strict" parsing
```
Comment (mjibson) [inline test/pgtest/copy-from-csv.pt:1]: Then this should go in the pgtest-mz directory.

File: test/pgtest/copy-from-csv.pt [hunk 5]
```diff
@@ -0,0 +1,139 @@
...
+DataRow {"fields":["7","seven\n8,eight\n"]}
+CommandComplete {"tag":"SELECT 7"}
+ReadyForQuery {"status":"I"}
+
+# Change options. Note that:
+# - After receiving terminating char, no other data should be accepted
+# - Quoted end of copy markers are still processed as end of copy markers
+send
+Query {"query": "DELETE FROM t"}
+Query {"query": "COPY t FROM STDIN WITH (FORMAT CSV, QUOTE '|', ESCAPE '#', NULL 'NS', HEADER true)"}
```
Comment (mjibson) [inline test/pgtest/copy-from-csv.pt:63]: Can you add tests with various combinations of delimiter, quote, and escape all being the same character?

**Review by mjibson — COMMENTED** [review]
Looking really good!

File: test/pgtest/copy-from-csv.pt (hunk 5 above)
Comment (sploiselle) [reply test/pgtest/copy-from-csv.pt]: Such a good call. Found out that the delimiter and quote character must differ. Added a test where delimiter and escape are the same character, though.

File: test/pgtest/copy-from-csv.pt (hunk 4 above)
Comment (sploiselle) [reply test/pgtest/copy-from-csv.pt]: Derp! tysm for catching that

File: src/pgwire/src/protocol.rs (hunk 3 above)
Comment (sploiselle) [reply src/pgwire/src/protocol.rs]: Didn't look as idiomatic to me but reverted to use Let

File: src/pgwire/src/message.rs (hunk 2 above)
Comment (sploiselle) [reply src/pgwire/src/message.rs]: Just an oversight; fixed

File: doc/user/content/sql/copy-from.md [hunk 6]
```diff
@@ -6,29 +6,66 @@ menu:
...
 
-Name | Default value | Description
+Name | Permitted values| Default value | Description
 -----|---------------|------------
-`DELIMITER` | tab character | Specifies the character that separates columns within each row (line) of the file.
-`NULL` | `\N` (backslash-N) | Specifies the string that represents a null value.
+`FORMAT` | `TEXT`, `CSV` | `TEXT` | Sets the input formatting method. For more information see [Text formatting](#text-formatting), [CSV formatting](#csv-formatting).
+`DELIMITER` | Single-quoted one-byte character | Format-dependent | Overrides the format's default column delimiter.
+`NULL` | Single-quoted strings | Format-dependent | Specifies the string that represents a _NULL_ value.
+`QUOTE` | Single-quoted one-byte character | `"` | Specifies the character to signal a quoted string, which may contain the `DELIMITER` value (without beginning new columns). To include the `QUOTE` character itself in column, wrap the column's value in the `QUOTE` character and prefix all instance of the value you want to literally interpret with the `ESCAPE` value. _`FROMAT CSV` only_
```
Comment (mjibson) [inline doc/user/content/sql/copy-from.md:27]: `FROMAT` typo

File: doc/user/content/sql/copy-from.md [hunk 7]
```diff
@@ -6,29 +6,66 @@ menu:
...
+`ESCAPE` | Single-quoted strings | `QUOTE`'s value | Specifies the character to allow instances of the `QUOTE` character to be parsed literally as part of a column's value. _`FROMAT CSV` only_
```
Comment (mjibson) [inline doc/user/content/sql/copy-from.md:28]: more fromat

File: test/pgtest-mz/copy-from-csv.pt [hunk 8]
```diff
@@ -0,0 +1,187 @@
...
+ErrorResponse {"fields":[{"typ":"C","value":"22P04"},{"typ":"M","value":"missing data for column"}]}
+ReadyForQuery {"status":"I"}
+CopyIn {"format":"text","column_formats":["text","text"]}
+ErrorResponse {"fields":[{"typ":"C","value":"22P04"},{"typ":"M","value":"extra data after last expected column"}]}
+ReadyForQuery {"status":"I"}
+
+send
+Query {"query": "COPY t FROM STDIN WITH (FORMAT CSV, DELIMITER '||')"}
+Query {"query": "COPY t FROM STDIN WITH (FORMAT CSV, QUOTE '||')"}
+Query {"query": "COPY t FROM STDIN WITH (FORMAT CSV, ESCAPE '||')"}
```
Comment (mjibson) [inline test/pgtest-mz/copy-from-csv.pt:150]: I think you need another non error case where quote and escape are the same, to trigger the double quote code path above.

File: doc/user/content/sql/copy-from.md (hunk 6 above)
Comment (sploiselle) [reply doc/user/content/sql/copy-from.md]: backticks stop my spellchecker from working! tysm

File: test/pgtest-mz/copy-from-csv.pt (hunk 8 above)
Comment (sploiselle) [reply test/pgtest-mz/copy-from-csv.pt:179]: Ah, I was glossing over that, but you're right that was missing added it and also having it test that the `ESCAPE` value defaults to the `QUOTE` value.

```
# ESCAPE defaults to QUOTE if not provided, and exhibits same behavior as
# default.
send
Query {"query": "DELETE FROM t"}
Query {"query": "COPY t FROM STDIN WITH (FORMAT CSV, QUOTE '#')"}
CopyData "1,#one#\n"
CopyData "2,###two###\n"
CopyData "3,#thr##ee#\n"
CopyDone
Query {"query": "SELECT * FROM t ORDER BY i"}
----

until
ReadyForQuery
ReadyForQuery
ReadyForQuery
----
CommandComplete {"tag":"DELETE 3"}
ReadyForQuery {"status":"I"}
CopyIn {"format":"text","column_formats":["text","text"]}
CommandComplete {"tag":"COPY 3"}
ReadyForQuery {"status":"I"}
RowDescription {"fields":[{"name":"i"},{"name":"t"}]}
DataRow {"fields":["1","one"]}
DataRow {"fields":["2","#two#"]}
DataRow {"fields":["3","thr#ee"]}
CommandComplete {"tag":"SELECT 3"}
ReadyForQuery {"status":"I"}
```

**Review by mjibson — APPROVED** [review]

**Comment by philip-stoev** [comment]
@sploiselle do you mind if I give this a try? If I do not get back with a verdict by the end of this week, please feel free to merge without waiting for me. Thanks!

**Comment by sploiselle** [comment]
@philip-stoev Of course--very glad to have additional tests. lmk what you find

**Pull request merged**